BeVolt/App/src/CLI.c

```diff
@@ -0,0 +1,723 @@
+	if(hashTokens[1] == NULL) {
+		if(contactor == ENABLE) {
+			printf("Contactor is Enabled\n\r");
+		} else {
+			printf("Contactor is Disabled\n\r");
+		}
+	} else {
+		printf("Invalid contactor command\n\r");
+	}
+}
+
+/** CLI_Charge
+ * Checks and displays the desired
+ * state of charge parameter(s)
+ */
+void CLI_Charge(void) {
+	if(hashTokens[1] == NULL) {
+		printf("The battery percentage is %.2f%%\n\r", SoC_GetPercent()/100.0);
+		return;
+	}
+	switch(hashTokens[1]) {
+		case RESET:
+			SoC_SetAccum(0);	//resets accumulator
+			printf("Accumulator has been reset\n\r");
+			break;
+		case SET:
+			SoC_SetAccum(hashTokens[2]);
+			printf("Accumulator has been set to %d%%\n\r", hashTokens[2]);
+			break;
+		default: 
+			printf("Invalid Charge Command\n\r");
+			break;
+	}
+}
+
+/** toggleLED
+ * Helper function for CLI_LED
+ * that toggles a given led
+ * @param led is the led to toggle
+ */
+void toggleLED(led input) {
+	if(hashTokens[2] == 1 || hashTokens[2] == ON) {
+		LED_On(input);
+	}
+	else if(hashTokens[2] == 0 || hashTokens[2] == OFF) {
+		LED_Off(input);
+	} else {
+		printf("Invalid LED command\n\r");
+	}
+}
+
+/** CLI_LED
+ * Interacts with the LEDs by 
+ * checking error light status
+ * running a full LED test
+ * and turning a specific LED on/off
+ */
+void CLI_LED(void) {
+	LED_Init();
+	uint8_t error = (GPIOB->ODR) & GPIO_Pin_12;
+	if(hashTokens[1] == NULL) {
+		if(error) {
+			printf("Error light is On\n\r");
+		} else {
+			printf("Error light is Off\n\r");
+		}
+		return;
+	}
+	switch(hashTokens[1]) {
+		case TEST:
+			printf("\n\r");
+			for(int i = 0; i < 10; i++) {
+				LED_Toggle(FAULT);
+				DelayMs(100);
+				LED_Toggle(RUN);
+				DelayMs(100);
+				LED_Toggle(UVOLT);
+				DelayMs(100);
+				LED_Toggle(OVOLT);
+				DelayMs(100);
+				LED_Toggle(OTEMP);
+				DelayMs(100);
+				LED_Toggle(OCURR);
+				DelayMs(100);
+				LED_Toggle(WDOG);
+				DelayMs(100);
+				LED_Toggle(CAN);
+				DelayMs(100);
+				LED_Toggle(EXTRA);
+				DelayMs(100);
+			}
+			break;
+		case FAULT_HASH:
+			toggleLED(FAULT);
+			break;
+		case RUN_HASH:
+			toggleLED(RUN);
+			break;
+		case OCURR_HASH:
+			toggleLED(OCURR);
+			break;
+		case OTEMP_HASH:
+			toggleLED(OTEMP);
+			break;
+		case OVOLT_HASH:
+			toggleLED(OVOLT);
+			break;
+		case WDOG_HASH:
+			toggleLED(WDOG);
+			break;
+		case CAN_HASH:
+			toggleLED(CAN);
+			break;
+		case EXTRA_HASH:
+			toggleLED(EXTRA);
+			break;
+		default:
+			printf("Invalid LED command\n\r");
+			break;
+	}
+}
+
+/** CLI_CAN
+ * Interacts with CAN by
+ * reading and writing to bus
+ */
+void CLI_CAN(void) {
+	uint8_t readData;
+	uint8_t writeData[8];
+	for(uint8_t i = 0; i < 7; i+= 2) {
+		char tempData[2] = {hexString[i], hexString[i+1]};
+		writeData[i/2] = strtol(tempData, NULL, 16);
+	}
+	switch(hashTokens[1]) {
+		case READ: 
+			while(CAN1_Read(&readData)){
+				printf("%d\n\r ", readData);
+			}
+		break;
+		case WRITE: 
+			switch(hashTokens[2]) {

```

Comment:
I don't like the switch statement stacking going on here, but I see why you did it.